File: bin/upload_curriculum.dart
// Usage: dart run bin/upload_curriculum.dart <filepath> [--force]
// Example: dart run bin/upload_curriculum.dart assets/courseConfigs/llpsi_familia_romana_v1.json
// Example: dart run bin/upload_curriculum.dart assets/courseConfigs/llpsi_familia_romana_v1.json --force

import 'dart:convert';
import 'dart:io';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_core/firebase_core.dart';
import 'firebase_options.dart';

void main(List<String> args) async {
  if (args.isEmpty) {
    print('Usage: dart run bin/upload_curriculum.dart <filepath> [--force]');
    print('Example: dart run bin/upload_curriculum.dart assets/courseConfigs/llpsi_familia_romana_v1.json');
    exit(1);
  }

  final filePath = args[0];
  final force = args.contains('--force');

  // Initialize Firebase
  await Firebase.initializeApp(options: DefaultFirebaseOptions.currentPlatform);
  final db = FirebaseFirestore.instance;

  try {
    // Read JSON file
    final file = File(filePath);
    if (!file.existsSync()) {
      print('❌ Error: File not found: $filePath');
      exit(1);
    }

    final jsonStr = file.readAsStringSync();
    final jsonData = jsonDecode(jsonStr) as Map<String, dynamic>;

    // Extract curriculum ID
    final configId = jsonData['id'] as String?;
    if (configId == null) {
      print('❌ Error: JSON must contain an "id" field');
      exit(1);
    }

    print('📖 Curriculum: $configId');
    print('📄 File: $filePath');

    // Check if document already exists
    final docRef = db.collection('courseConfigs').doc(configId);
    final docSnap = await docRef.get();

    if (docSnap.exists && !force) {
      print('⚠️  Document already exists. Use --force to overwrite.');
      print('   Command: dart run bin/upload_curriculum.dart $filePath --force');
      exit(1);
    }

    if (docSnap.exists && force) {
      print('⚠️  Overwriting existing document...');
    }

    // Add metadata fields
    jsonData['active'] = true;
    jsonData['uploadedAt'] = FieldValue.serverTimestamp();
    jsonData['updatedAt'] = FieldValue.serverTimestamp();

    // Upload to Firestore
    await docRef.set(jsonData, SetOptions(merge: false));

    print('✅ Successfully uploaded to Firestore!');
    print('   Collection: courseConfigs');
    print('   Document ID: $configId');
    exit(0);
  } catch (e) {
    print('❌ Error: $e');
    exit(1);
  }
}

File: scripts/upload-curriculum.js
#!/usr/bin/env node

/**
 * Upload curriculum JSON to Firestore
 * Usage: node scripts/upload-curriculum.js <filepath> [--force]
 * Example: node scripts/upload-curriculum.js assets/courseConfigs/llpsi_familia_romana_v1.json
 */

const fs = require('fs');
const path = require('path');
const admin = require('firebase-admin');

// Initialize Firebase Admin SDK
const serviceAccountPath = process.env.FIREBASE_SERVICE_ACCOUNT || 
  path.join(__dirname, '..', 'firebase-service-account.json');

if (!fs.existsSync(serviceAccountPath)) {
  console.error('❌ Error: Firebase service account not found at:', serviceAccountPath);
  console.error('   Set FIREBASE_SERVICE_ACCOUNT environment variable or place json file at project root');
  process.exit(1);
}

const serviceAccount = require(serviceAccountPath);
admin.initializeApp({
  credential: admin.credential.cert(serviceAccount),
});

const db = admin.firestore();

async function uploadCurriculum(filePath, force = false) {
  try {
    // Read and parse JSON
    if (!fs.existsSync(filePath)) {
      console.error(`❌ Error: File not found: ${filePath}`);
      process.exit(1);
    }

    const jsonStr = fs.readFileSync(filePath, 'utf-8');
    const curriculum = JSON.parse(jsonStr);

    // Validate
    const configId = curriculum.id;
    if (!configId) {
      console.error('❌ Error: JSON must contain an "id" field');
      process.exit(1);
    }

    console.log('📖 Curriculum:', configId);
    console.log('📄 File:', filePath);

    // Check if exists
    const docRef = db.collection('courseConfigs').doc(configId);
    const docSnap = await docRef.get();

    if (docSnap.exists && !force) {
      console.error('⚠️  Document already exists. Use --force to overwrite.');
      console.error('   Command: node scripts/upload-curriculum.js', filePath, '--force');
      process.exit(1);
    }

    if (docSnap.exists && force) {
      console.log('⚠️  Overwriting existing document...');
    }

    // Add metadata
    curriculum.active = true;
    curriculum.uploadedAt = admin.firestore.FieldValue.serverTimestamp();
    curriculum.updatedAt = admin.firestore.FieldValue.serverTimestamp();

    // Upload
    await docRef.set(curriculum, { merge: false });

    console.log('✅ Successfully uploaded to Firestore!');
    console.log('   Collection: courseConfigs');
    console.log('   Document ID:', configId);
    
    process.exit(0);
  } catch (err) {
    console.error('❌ Error:', err.message);
    process.exit(1);
  }
}

// Parse arguments
const args = process.argv.slice(2);
if (args.length === 0) {
  console.log('Usage: node scripts/upload-curriculum.js <filepath> [--force]');
  console.log('Example: node scripts/upload-curriculum.js assets/courseConfigs/llpsi_familia_romana_v1.json');
  process.exit(1);
}

const filePath = args[0];
const force = args.includes('--force');

uploadCurriculum(filePath, force);
